src/VisualStudio/Core/Def/Implementation/LanguageClient/RazorInProcLanguageClient.cs

```diff
@@ -49,10 +52,15 @@ internal class RazorInProcLanguageClient : AbstractInProcLanguageClient
             DefaultCapabilitiesProvider defaultCapabilitiesProvider)
             : base(languageServerProtocol, workspace, diagnosticService, listenerProvider, solutionProvider, ClientName)
         {
+            _globalOptionService = globalOptionService;
             _defaultCapabilitiesProvider = defaultCapabilitiesProvider;
         }
 
         protected internal override VSServerCapabilities GetCapabilities()
-            => _defaultCapabilitiesProvider.GetCapabilities();
+        {
+            var capabilities = _defaultCapabilitiesProvider.GetCapabilities();
+            capabilities.SupportsDiagnosticRequests = _globalOptionService.GetOption(InternalDiagnosticsOptions.RazorDiagnosticMode) == DiagnosticMode.Pull;
+            return capabilities;
```

Comment:
this part is new.  Here's how Razor checks the option to opt-into pull diags for their language client.